### PR TITLE
fix(web): proper linkage of sources to events 🪠

### PR DESCRIPTION
**Pull request by jahorton**
Continues from #10843. 

Despite my efforts to avoid it, it appears necessary to retool the event engines to ensure a proper linkage between original event identifier and the corresponding `GestureSource` that will be generated and updated.  The previous "identifier" to "source" mapping style appears to be too "loose" due to asynchronicity - it's time to update to a strategy that can directly use closure-capture mechanics to ensure proper binding.

**Extra details**:  By default, browsers will reuse touch identifiers after they are freed - identifiers aren't unique.  If this replacement happens while related events are deferred, we get problems like what can be seen in #10843's user test reports.  Thus, it's necessary to establish the link between identifier and source while the identifier is _still_ current for the touchpath represented by the source, then "lock in" that link - something that closure-capturing is well-suited for.  

As the actual construction is queued for later, the best way I can see to leverage closure-capture mechanisms to resolve the issue is to creating Promises for the `GestureSource` synchronously during the actual touchStart handler - not within its closures.  _Resolution_ of the `Promise`s can occur within closures, but assignments leveraging the `Promise` need to be maintained synchronously, as it can then be accessed synchronously within the other events for closure-capture in their handlers.  (After all, under extreme rapid-typing situations, it's quite possible the `Promise` will only be fulfilled long after the fact, but this Promise will always be available, unambiguously, in time with the original events.)

## User Testing

**TEST_RAPID_TYPING_STABILITY**:  Using the Keyman for Android app, attempt to reproduce #10592 and #10646.

1. Set "EuroLatin (SIL)" as the active keyboard.
2. Rapidly type 10 random keys.  Verify that 10 characters are emitted, possibly after a brief wait.
    - Don't worry about accuracy - just try to type 10 characters as fast as humanly possible, whatever they are.
    - Keep in mind that a space (` `) does count as a character; it's easy to accidentally hit it when not worried about accuracy.
3. Rapidly type while utilizing functions like the globe key, space bar and predictive texts intermittently.  (Refer to https://github.com/keymanapp/keyman/pull/10641#issuecomment-1935882421 if needed.)
4. Try to rapidly type something specific while being precise.
    - Perhaps `typing rapidly should not skip keys`; just make sure it's something you feel comfortable typing rapidly.
    - If any of the letters appear out-of-order or appear to be skipped entirely, FAIL this test.
    - Do not fail this test if you suspect you simply fat-fingered a key, hitting a close neighbor instead of the intended key.
5. FAIL this test if any of the following happens:
    - Stuck key-highlighting
    - Stuck key-previews
    - Stuck subkey menu
    - Stuck modipress

**TEST_GESTURE_REGRESSIONS**: Run the full set of Web's gesture-related regression tests and report back any errors.
- For the longterm, it would likely be best to add `TEST_RAPID_TYPING_STABILITY` to that set of tests.
- I assume you have a copy of [these tests](https://github.com/keymanapp/keyman/wiki/Platform-Acceptance-Test-for-Web#suite_touch_gestures-gesture-oriented-touch-platform-tests-new-in-170-from-feature-gestures) done up in your preferred format; if not, copy and paste the ones from [SUITE_TOUCH_GESTURES](https://github.com/keymanapp/keyman/wiki/Platform-Acceptance-Test-for-Web#suite_touch_gestures-gesture-oriented-touch-platform-tests-new-in-170-from-feature-gestures) here.

### CODE REVIEW

File: common/web/gesture-recognizer/src/engine/touchEventEngine.ts [hunk 1]
```diff
@@ -185,13 +232,34 @@ export class TouchEventEngine<HoveredItemType, StateToken = any> extends InputEv
...
 
+    /*
+      Using the Promise map built in touchStart, we can retrieve a Promise for the source linked
+      to this event and closure-capture it for the closure queued below.
+    */
+    const capturedSourcePromises = new Map<number, Promise<GestureSource<ItemType, StateToken>>>();
+    for(let i = 0; i < event.touches.length; i++) {
+      const touchId = event.touches.item(i).identifier;
+      capturedSourcePromises.set(touchId, this.pendingSourcePromises.get(touchId).corePromise);
+    }
```
Comment (jahorton) [inline common/web/gesture-recognizer/src/engine/touchEventEngine.ts:243]: The big difference from #10843 is that we can closure-capture the Promises at the original time of the event, unlike the `GestureSource` instances.  #10843 tried to make the mapping work asynchronously, but there were cases that arose in which the former, more loosely-bound style simply couldn't work consistently.

That was likely because the mapping could see an entry erased by a different `GestureSource` (conceptually) that was reusing the same identifier.

**Comment by bharanidharanj** [comment]
# Test Results

* **TEST_RAPID_TYPING_STABILITY (FAILED):** Tested with the attached PR build (Keyman 17.0.271-beta-test-10960) on an Android 13 Mobile device and here is my observation: 1. Opened the Keyman In-app. 2. Rapidly typing on OSK leads to stuck-key-highlighting issues. 3. Tested the same on an Android 14 / API 34 emulator and I was able to reproduce it.

File: common/web/gesture-recognizer/src/engine/touchEventEngine.ts [hunk 2]
```diff
@@ -115,49 +114,104 @@ export class TouchEventEngine<HoveredItemType, StateToken = any> extends InputEv
...
+      const touch = event.changedTouches.item(i);
+      const promise = new ManagedPromise<GestureSource<ItemType, StateToken>>();
+      this.pendingSourcePromises.set(touch.identifier, promise);
+      capturedSourcePromises.set(touch.identifier, promise);
+    }
+
+    /*
+      When multiple touchpoints are active, we need to ensure a specific order of events.
+      The easiest way to ensure the exact order involves programmatic delay of their
+      processing, essentially "sequentializing" the events into a deterministic order.
```
Comment (mcdurdin) [inline common/web/gesture-recognizer/src/engine/touchEventEngine.ts:155]: Is this delay visible to the end user?

**Review by mcdurdin — COMMENTED** [review]
I understand why you are making these changes and in principle they seem fine. This is a significant change to the structure of the gesture engine for beta, however, so we will definitely need to make sure we do extra testing for stability and unforeseen edge cases.

The only user-visible impact (apart from fixing instability, which is obviously good) seems to be the introduction of a delay? That seems less than ideal -- can you describe the impact?

**Comment by jahorton** [comment]
I didn't see anything about an Android error this time, which is a plus.  Also... turns out that I made a fairly simple mistake in one of the event handlers - I forgot that ending touches don't show up in their event's `.touches` property - only within the `.changedTouches` property.  Our unit-test event simulation didn't meet that aspect of the spec and so failed to catch this detail.

@keymanapp-test-bot retest ALL

File: common/web/gesture-recognizer/src/engine/touchEventEngine.ts (hunk 2 above)
Comment (jahorton) [reply common/web/gesture-recognizer/src/engine/touchEventEngine.ts:155]: It should be imperceptible; it's a matter of queuing tasks against the microtask queue, the macrotask queue, and the custom queue from #10843.  Each of those should take negligible amounts of time.

The only delay added in this PR (at present) is due to the need to await a few pre-resolved Promises... so a couple of microtask delays per queued event should be the only extra wait here.  (They're not resolved when closure-captured, but they should already be resolved at every point they're being `await`-ed.)

Comment (jahorton) [reply common/web/gesture-recognizer/src/engine/touchEventEngine.ts:155]: If I spent some time to dwell on things, perhaps I could 'rig' a way to avoid using Promises and the microtask queue for this, which would remove even the microtask queue waits... but that might actually make the code that little bit more difficult to parse for maintenance.  I feel that `Promise` semantics work in our favor here.

A simple object (providing a closure-capturable root reference) with a mutatable field would be enough.  Said field would start off `undefined` but be assigned the source's reference when it's available (during the final `onTouchStart` closure).  Of course, we'd have to explicitly document and maintain that connection, which is why I feel it a bit less "simple."

Comment (mcdurdin) [reply common/web/gesture-recognizer/src/engine/touchEventEngine.ts:155]: My preference is almost always for synchronous code where possible -- re-entrant code makes code flow more complicated with potential for other interactions (from other promises and events) which all goes away without it. But I'd need to compare the two solutions here to be sure.

**Comment by bharanidharanj** [comment]
# Test Results

* **TEST_RAPID_TYPING_STABILITY (PASSED):** Retested with the updated PR build (Keyman 17.0.285-beta-test-10960) in various Android OS versions and here is my observation: 1. Followed the steps as mentioned in the test description and I did not find any error message in Android emulators (Android 5.0, Android 7.1, Android 9.0 and Android 14.0). 2. I did not find any stuck-key-highlighting or any other issues (eg., stuck key-previews, stuck subkey menu, and stuck modipress)  during the process. 3. FYI, I got a Keyboard error message while testing it in an Android 13 Mobile device. Apart from this, this build is working fine. 

..Android 5.0 / API 21 emulator

   


..Android 7.1.1 / API 25 emulator

   


..Android 9.0 / API 28 emulator

   


..Android 14.0 / API 34 emulator 

   


..Android 13.0 / Samsung Galaxy A23 Mobile device

**Comment by jahorton** [comment]
About those two Sentry reports:  yay, the old error types are dead, and it's something new now - and it's _much_ simpler to resolve _and_ to reason about than the old patterns.

I can also see why the user experience wasn't affected by the errors that arose; the new pattern makes it clear that those errors are triggered from already-completed gestures.  Still, it's best if we don't even have errors to report, yeah?

@keymanapp-test-bot retest TEST_RAPID_TYPING_STABILITY 

One more time - let's aim for zero Sentry error reports this round.

**Comment by jahorton** [comment]
> [...] after typing rapidly using the OSK for a long time. (approx. 4 minutes) [...]. 
> Seems to be working fine.

Well... if nothing else, that's certainly quite rare, and is a massive improvement on what we previously had.

Looking at Sentry, I'm not seeing any new error reports - the only stuff I see is from the previous user-test run.  Nothing new in Web or under Android within the `test` environment sets.

**Comment by mcdurdin** [comment]
Yes it's an improvement, but a keyboard that crashes every 4 minutes is not exactly stable.

**Comment by jahorton** [comment]
> Yes it's an improvement, but a keyboard that crashes every 4 minutes is not exactly stable.

I agree... though it's hard to know what to address at the moment.  Unlike prior user test runs, I'm not seeing _any_ Sentry events to follow up on.  Kinda need those to better address things and find the root sources.

I _think_ sometimes error reports might have been withheld after one user test run and "released" somehow on the next?  There have been times where I went looking for Sentry events and didn't see them... but after a user-retest, errors that should've been there to begin with were suddenly visible.  I'm still not 100% sure on how or why that was happening. 

Maybe, maybe rerunning the test for a bit might allow error reports through.  This is just a feeling / suspicion though, and isn't a guarantee.  It's certainly bugging me that we're getting error toasts that don't send Sentry error reports to match.

**Review by mcdurdin — APPROVED** [review]
I think I understand all the changes now. I'm not sure why KMW is 3KB larger -- there's not that much new code?

**Comment by bharanidharanj** [comment]
# Test Results

* **TEST_GESTURE_REGRESSIONS (FAILED):** Run the Gesture_Regression test on the following browsers and setups, and here are my observations: 

1. Android_Chrome_touch: Tested on an Android device (version 13) using the Chrome browser. The test failed in 7 baseline tests.
Android Test failed in 
Test_Basic_Modepress_Hold
 Tes_Numeric_From_Shift
Test_Delayed_subkey
Test_Custom_Multitap_Modifier
Test_Alternating_Shift_And_Key
Test_Modpress_Multitap_Flic_Preview
Test_Flick_During_Modpress

..Test_Alternating_Shift_and_Key


https://github.com/keymanapp/keyman/assets/19683143/fa884651-dac8-4a29-8132-a87bb115726a


2. Tested on an iPhone 13 mobile device (iOS version 17.4) using the Safari browser, and all tests appear to have passed. A sample screenshot is attached for reference.

..Flick_Locking

   

3. Tested on an iPad (9th generation) using the iOS 15.5 simulator with the Safari browser. I observed that certain tests couldn't be performed in the simulator because they require two actions (holding the shift key and then long-pressing another key).

The following tests failed due to the previously mentioned issue.
Test_Basic_Modpress
Test_Basic_Modepress_Hold
Test_Numeric_From_Shift
Test_Delayed_subkey
Test_Custom_Multitap_Modifier
Test_Alternating_Shift_And_Key
Test_Modpress_Multitap_Flick_Preview
Test_Flick_During_Modpress
Test_Flick_Basics
Test_Flick_Correction
Test_Flick_Locking

**Comment by jahorton** [comment]
I've been noticing a few gesture things along the lines of the failed test recently too on my personal iPhone... but hadn't realized it was happening that extensively.  Interestingly, I'm not seeing some of the issues in iOS Simulator.

Did some digging via TestFlight and found that 17.0.261-beta seems to be comparatively fine, with a number of breaks happening in 17.0.262-beta.  Interestingly... there's no obvious change that should have landed then to trigger the issue.

**Comment by mcdurdin** [comment]
> "Certain calculations" - we rely on key layout information from the DOM when noting the 'item' for each ongoing touchpath. To get this for layers currently not visible, we _make_ them visible temporarily, which triggers layout reflows - and _that_ appears to be a major contributing factor to lagginess on the local test device.

Hmm, could we cache this information when the keyboard is first loaded? (Perhaps being smart enough to think about rotation as well?) This may not be a 17.0-beta change but I'm all for avoiding layout reflows during processing!

**Comment by jahorton** [comment]
In regard to that last point...

image

Drilling down the other high-percentage contributors reveals similar figures - a lot of time was spent on layout-reflow.  I may want to spend some time seeing if that can be optimized.

**Comment by mcdurdin** [comment]
> Drilling down the other high-percentage contributors reveals similar figures - a lot of time was spent on layout-reflow. I may want to spend some time seeing if that can be optimized.

Sounds good to me! Let's make sure we get the other beta PRs merged before you get too deep into this one though.

> Interestingly, I don't see anything even _close_ to this level of lagginess within the iOS app/keyboard; maybe Safari webviews are caching layout calculations in a manner that Chrome webviews aren't?

Potentially -- but also iPhones are much faster hardware so it may just be much less visible.

**Comment by jahorton** [comment]
> > "Certain calculations" - we rely on key layout information from the DOM when noting the 'item' for each ongoing touchpath. To get this for layers currently not visible, we _make_ them visible temporarily, which triggers layout reflows - and _that_ appears to be a major contributing factor to lagginess on the local test device.
> 
> Hmm, could we cache this information when the keyboard is first loaded? (Perhaps being smart enough to think about rotation as well?) This may not be a 17.0-beta change but I'm all for avoiding layout reflows during processing!

I could see caching each layer's layout info as it's loaded.  That would be well within reason.  (Then dumping the cache on a rotation or keyboard size-shift.)

**Comment by jahorton** [comment]
> > Drilling down the other high-percentage contributors reveals similar figures - a lot of time was spent on layout-reflow. I may want to spend some time seeing if that can be optimized.
> 
> Sounds good to me! Let's make sure we get the other beta PRs merged before you get too deep into this one though.

Agreed.  If this is the only real source of problems at present, that can certainly be done _separately_, allowing this chain to merge.

**Comment by mcdurdin** [comment]
Just wondering: can we not calculate the regions without doing layout at all? We have all the numbers in the touch layout data and we have the bounding box size?

**Comment by jahorton** [comment]
> Just wondering: can we not calculate the regions without doing layout at all? We have all the numbers in the touch layout data and we have the bounding box size?

At the very least, we can approximate it with that data - and we might even be able to perfectly match it.  Sounds like a wise optimization path to investigate.

**Comment by jahorton** [comment]
> I'm not sure why KMW is 3KB larger -- there's not that much new code?

This is 6 PRs in on the 🪠 chain.  This is the one that pushes things over on the filesize warning, I guess.

**Comment by bharanidharanj** [comment]
# Test Results

* **TEST_GESTURE_REGRESSIONS (FAILED):** Retested using the 'Test uniminified KeymanWeb' test page on an Android (13) mobile device and here is my observation: 1. Followed the instructions as per in the Suite_Touch_Gestures page. 2. I did not see any issue up to Test_Basic_Modipress_Hold. 3. There is a sluggishness on the keyboard after longpressing % key (step 6) and then releasing it. Also, noticed that the keyboard changes back to the default layer. Seems to be an issue. I have attached the video file for reference. 

https://github.com/keymanapp/keyman/assets/19683143/8b03ab00-1618-4ca8-bdf2-d00d90578dbb

**Comment by jahorton** [comment]
The remaining issues seem to mostly be performance-related.  There is one notable, though ambiguous, functionality point of failure... but this does resolve a _lot_ of far, far worse behaviors and will let us get `beta` to a better state.  As such, @mcdurdin and I have made the decision to go ahead and merge this PR and its predecessors.  (The next one in line is still pending review.)

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 17.0.301-beta](https://keyman.com/downloads/releases/beta/17.0.301)
